lib/master/worker_manager.go

```diff
@@ -0,0 +1,494 @@
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	allPersistedWorkers, err := m.workerMetaClient.LoadAllWorkers(ctx)
+	if err != nil {
+		return err
+	}
+
+	for workerID, status := range allPersistedWorkers {
+		if entry, exists := m.workerEntries[workerID]; !exists {
+			// For those workers we have lost contact with, we insert
+			// a tombstone.
+			m.workerEntries[workerID] = newTombstoneWorkerEntry(workerID, status)
+		} else {
+			// Put the current persisted status in the workerEntry.
+			entry.InitStatus(status)
+		}
+	}
+
+	m.state = workerManagerReady
+	return nil
+}
+
+func (m *WorkerManager) HandleHeartbeat(msg *libModel.HeartbeatPingMessage, fromNode p2p.NodeID) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if !m.checkMasterEpochMatch(msg.Epoch) {
+		return nil
+	}
+
+	entry, exists := m.workerEntries[msg.FromWorkerID]
+	if !exists {
+		if m.state != workerManagerWaitingHeartbeat {
+			log.L().Info("Message from stale worker dropped",
+				zap.String("master-id", m.masterID),
+				zap.Any("message", msg),
+				zap.String("from-node", fromNode))
+			return nil
+		}
+
+		// We are expecting heartbeats now
+		entry = newWorkerEntry(
+			msg.FromWorkerID,
+			model.ExecutorID(fromNode),
+			m.nextExpireTime(),
+			false,
+			nil)
+		m.workerEntries[msg.FromWorkerID] = entry
+
+		log.L().Info("Worker discovered", zap.String("master-id", m.masterID),
+			zap.Any("worker-entry", entry))
+
+		entry.heartbeatCount.Store(1)
+		// We do not call onWorkerOnlined because the master is still waiting to be contacted.
+		return nil
+	}
+
+	entry.SetExpireTime(m.nextExpireTime())
+
+	newHeartbeatCount := entry.heartbeatCount.Add(1)
+	if newHeartbeatCount == 1 {
+		err := m.enqueueEvent(&masterEvent{
+			Tp:       workerOnlineEvent,
+			WorkerID: msg.FromWorkerID,
+			Handle: &runningHandleImpl{
+				workerID:   msg.FromWorkerID,
+				executorID: model.ExecutorID(fromNode),
+				manager:    m,
+			},
+		})
+		if err != nil {
+			return nil
+		}
+	}
+	return nil
+}
+
+// Tick should be called by the BaseMaster so that the callbacks can be
+// run in the main goroutine.
+func (m *WorkerManager) Tick(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	for {
+		var event *masterEvent
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case event = <-m.eventQueue:
+		default:
+			return nil
+		}
+
+		switch event.Tp {
+		case workerOnlineEvent:
+			if err := m.onWorkerOnlined(ctx, event.Handle); err != nil {
+				return err
+			}
+		case workerOfflineEvent:
+			if err := m.onWorkerOfflined(ctx, event.Handle, event.Err); err != nil {
+				return err
+			}
+		case workerStatusUpdatedEvent:
+			if err := m.onWorkerStatusUpdated(ctx, event.Handle); err != nil {
+				return err
+			}
+		case workerDispatched:
+			if err := m.onWorkerDispatched(ctx, event.Handle, event.Err); err != nil {
+				return err
+			}
+		}
+
+		if event.beforeHook != nil {
+			event.beforeHook()
+		}
```

Comment:
Could you please explain when naming it `beforeHook`?

Reply:
It is a function called before the main callback is called.